rfc/rfc-0000-simplify-runnable-use.md

```diff
@@ -0,0 +1,391 @@
+              value: $(runnable.spec.inputs.repository)$
+            - name: revision
+              value: $(runnable.spec.inputs.revision)$
+         workspaces:
+            - name: shared-workspace
+              volumeClaimTemplate:
+                 spec:
+                    accessModes:
+                       - ReadWriteOnce
+                    resources:
+                       requests:
+                          storage: 256Mi
+   outputs:
+      url: spec.params[?(@.name=="repository")].value
+      revision: spec.params[?(@.name=="revision")].value
+
+---
+apiVersion: carto.run/v1alpha1
+kind: ClusterSourceTemplate
+metadata:
+   name: source-linter
+spec:
+   template:
+      apiVersion: carto.run/v1alpha1
+      kind: Runnable
+      metadata:
+         name: $(workload.metadata.name)$-linter
+      spec:
+         runTemplateRef:
+            name: md-linting-pipelinerun
+         inputs:
+            repository: $(workload.spec.source.git.url)$
+            revision: $(workload.spec.source.git.revision)$
+         serviceAccountName: pipeline-run-management-sa
+   urlPath: .status.outputs.url
+   revisionPath: .status.outputs.revision
+```
+
+## Proposed example
+Under this proposal, users would only specify one object with a new immutable
+flag.
+
+```yaml
+apiVersion: carto.run/v1alpha1
+kind: ClusterSourceTemplate    # <== Now a source template
+metadata:
+   name: source-linter
+spec:
+   template:
+      apiVersion: tekton.dev/v1beta1
+      kind: PipelineRun
+      metadata:
+         name: $(workload.metadata.name)$-linter      # <== Refers to workload
+      spec:
+         pipelineRef:
+            name: linter-pipeline
+         params:
+            - name: repository
+              value: $(workload.spec.source.git.url)$      # <== Refers to workload
+            - name: revision
+              value: $(workload.spec.source.git.revision)$ # <== Refers to workload
+         workspaces:
+            - name: shared-workspace
+              volumeClaimTemplate:
+                 spec:
+                    accessModes:
+                       - ReadWriteOnce
+                    resources:
+                       requests:
+                          storage: 256Mi
+   urlPath: spec.params[?(@.name=="repository")].value    # <== Refers to path on the immutable object that will be created
+   revisionPath: spec.params[?(@.name=="revision")].value
+   lifecycle: immutable                                         # <== Flag
+```
+
+# How it Works
+[how-it-works]: #how-it-works
+
+When Cartographer processes a template with the immutable flag, rather than
+creating/updating the templated object it will create/update a Runnable and
+ensure that a matching ClusterRunTemplate exists.
+
+What is meant by "ensure that a matching ClusterRunTemplate exists"? For any
+given supply chain template (a cluster scoped object) there will only need
+to be one ClusterRunTemplate. All of the N Runnables created by N workloads
```

Comment:
I wonder if we're getting into a non crucial optimization here - as ClusterRunTemplate objects just carry data, I wonder if it'd be enough to just always create one for a given runnable (then be able to rely on `generateName` for these). wdyt?

Reply:
For every ClusterSourceTemplate that stamps out an immutable object, there is one logical ClusterRunTemplate. That ClusterRunTemplate is determinative (it uses no information from the workload). I don't think we should create N ClusterRunTemplates which are all identical except for their name, when instead we could create just one.

Reply:
Ensuring that an expected object exists is also standard Carto behavior. For every workload, the controller ensures that the expected objects have been created.

Reply:
In some discussions the concern for a race condition has been raised. I do not believe that is a concern. In the worst case scenario a user defines a ClusterSourceTemplate with an immutable lifecycle. This template is then referenced by both a delivery and a supply chain. A reconciliation is kicked off for both at the same time. Both check for the existence of the expected ClusterRunTemplate. Neither see it. Both submit the same ClusterRunTemplate with the same name (because it is based on the name of the ClusterSourceTemplate). One submission is handled first and the other second. No conflict exists.

I welcome further poking of this, but I do not see a concern about name collisions or race conditions.